File: weibo.cn/cpp/spliter/spliter.cpp
#include "spliter.h"


Spliter::Spliter(std::string save_dir)
{
    this->save_dir = save_dir;
}

void Spliter::split_letters(std::string filename, cv::Mat letters[4])
{

    //read the image
    cv::Mat image =  cv::imread (filename, CV_LOAD_IMAGE_COLOR);
    cv::imshow ("1", image);

    // clean noise line
    this->clean_noise(image);

    //split the image
    int splits[8];
    captcha_utils.vertical_project (image, splits);
//    cv::Mat clone_mat = image.clone();
//    for(int i = 0; i < 8; i++){
//        std::cout<<i<<" "<<splits[i]<<std::endl;
//        cv::line (clone_mat, cv::Point(splits[i], 0), cv::Point(splits[i], 20), cv::Scalar(0));
//    }
//    cv::imshow ("split", clone_mat);

    if(splits[7] < image.size().width){//assume the final split is in the range of image.
        for(int i = 0; i < 8; i += 2){
            cv::Rect rect(cv::Point(splits[i], 0), cv::Point(splits[i+1], 20));
            letters[i / 2] = image(rect).clone();
        }
    }
}

void Spliter::split_and_save(std::string filename)
{
    cv::Mat letters[4]; 
    this->split_letters(filename, letters);
    for (int i = 0; i < 4; ++i){
        this->save_image(letters[i]);
    }
}

void Spliter::save_image(cv::Mat &splited_mat)
{
    if(splited_mat.size().width > 32) return;

    //put the mat in center of 32*32 pixels image.
    int out_height = 32;
    int out_width = 32;
    cv::Mat out(out_height, out_width, CV_8UC1,cv::Scalar(255));
    int offset_x = (out_width - splited_mat.size().width)/2;
    int offset_y = (out_height - 20)/2;
    cv::Mat imageROI;
    imageROI = out(cv::Rect(offset_x, offset_y, splited_mat.cols, splited_mat.rows));
    splited_mat.copyTo(imageROI);

    //generate random filename
    boost::uuids::random_generator rgen;
    boost::uuids::uuid ranUUID = rgen();
    std::string filename = this->save_dir + boost::lexical_cast<std::string>(ranUUID) + ".png";
    cv::imwrite(filename, out);
}


bool is_black(int i, int j, cv::Mat &image);
void clear_color(cv::Mat& image);
int get_horizontal_noise_line_width(cv::Mat &image, int now_height, int now_width);
void clear_horizontal_noise_line(cv::Mat &image);

void Spliter::clean_noise(cv::Mat &image)
{
    //clean noise line
    cv::flip (image, image, -1);
    cv::imshow("2", image);
    clear_horizontal_noise_line(image);
    cv::imshow("3", image);
    cv::waitKey(0);
    cv::flip (image, image, -1);
    clear_horizontal_noise_line(image);
    clear_color(image);
//    cv::imshow("clear", image);

    //threshold the image
    cv::cvtColor(image, image, CV_BGR2GRAY);//BGR to Gray
    cv::threshold (image, image, 150, 255, cv::THRESH_BINARY);
//    cv::imshow ("thresholded", image);

    //clean peper noise
    captcha_utils.clear_peper_noise(image, 2);
//    cv::imshow("clear peper:", image);
}

bool is_black(int i, int j, cv::Mat &image)
{
    int b =  image.at<cv::Vec3b>(j, i)[0];
    int g =  image.at<cv::Vec3b>(j, i)[1];
    int r =  image.at<cv::Vec3b>(j, i)[2];
    int average = (r + g + b)/3;
    if( r <244 &&  (abs(average-b) < 4)
        && (abs(average-g) < 4)
        && (abs(average-r) < 4)){
        return true;
    }
    return false;
}


void clear_color(cv::Mat& image)
{
    for(int i = 0; i < image.size().width; i++){
        for(int j = 0; j < image.size().height; j++){
           if(is_black(i, j, image)){
               image.at<cv::Vec3b>(j, i)[0] = 20;
               image.at<cv::Vec3b>(j, i)[1] = 20;
               image.at<cv::Vec3b>(j, i)[2] = 20;
           }
        }
    }
}


int get_horizontal_noise_line_width(cv::Mat &image, int now_height, int now_width)
{
    using namespace std;
    int end_width = now_width;
    while(end_width < image.size().width
          && image.at<cv::Vec3b>(now_height, end_width)[0] < 12
          && image.at<cv::Vec3b>(now_height, end_width)[1] < 12
          && image.at<cv::Vec3b>(now_height, end_width)[2] < 12){
        
        // cout<<
        //   int(image.at<cv::Vec3b>(now_height, end_width)[0])<< " "<<
        //   int(image.at<cv::Vec3b>(now_height, end_width)[1])<< " "<<
        //   int(image.at<cv::Vec3b>(now_height, end_width)[2])<< " "<<endl;
        end_width++;
    }
    return end_width - now_width;
}



void clear_horizontal_noise_line(cv::Mat &image)
{
    using namespace std;
    int first_height;
    bool has_find = false;
    for(int i = 0; i < image.size().height; i++){
        //if three continuous pixels are black horizontally， 
        //   these three pixels are part of noise line
        if(image.at<cv::Vec3b>(i, 0)[0] < 12
                && image.at<cv::Vec3b>(i, 0)[1] < 12
                && image.at<cv::Vec3b>(i, 0)[2] < 12
                && get_horizontal_noise_line_width(image, i, 0) >= 2 ){
            first_height = i;
            has_find = true;
        }
    }
    if(!has_find) return;
    cout<<"first:"<<first_height<<endl;
    exit(0);
    int now_width = 0;
    int now_height = first_height;
    while(now_width < image.size().width){
        int width = get_horizontal_noise_line_width(image, now_height, now_width);
        cout<<now_width<<"  "<<now_height<< " width:"<<width<<endl;
        //clear the horizontal noise line
        for(int i = now_width; i < now_width + width; i++){
            int top_num = 0;
            int bottom_num = 0;
            //the upper pixel
            if(is_black(i, now_height-1, image)) top_num++;
            //the upper left pixel 
            if(is_black(i-1, now_height-1, image)) top_num++;
            //the upper right pixel
            if(is_black(i+1, now_height-1, image)) top_num++;
            //the lower pixel
            if(is_black(i, now_height+1, image)) bottom_num++;
            //the left lower pixel
            if(is_black(i-1, now_height+1, image)) bottom_num++;
            //the right lower pixel
            if(is_black(i+1, now_height+1, image)) bottom_num++;

            if(now_height != 0 && now_height != image.size().height){
                if(top_num>0 && bottom_num>0) continue;
            }

            image.at<cv::Vec3b>(now_height, i)[0] = 255;
            image.at<cv::Vec3b>(now_height, i)[1] = 255;
            image.at<cv::Vec3b>(now_height, i)[2] = 255;
        }

        //find the next noise pixel
        int a = get_horizontal_noise_line_width(image, now_height - 1, now_width + width -1);
        int b = get_horizontal_noise_line_width(image, now_height + 1, now_width + width -1);
        int c = get_horizontal_noise_line_width(image, now_height - 1, now_width + width);
        int d = get_horizontal_noise_line_width(image, now_height + 1, now_width + width);
        if(now_height == 0) a=0,c=0;
        if(now_height == (image.size().height - 1) ) b=0,d=0;
        //cout<<"abcd: "<<a<<" "<<b<<" "<<c<<" "<<d<<endl;

        int max_a_b =  a>b?a:b;
        int max_c_d =  c>d?c:d;
        int max_a_b_c_d = max_a_b>max_c_d?max_a_b:max_c_d;
        //cout<<"max_abcd:"<<max_a_b_c_d<<endl;
        if(max_a_b_c_d < 2) break;
        if(max_a_b == max_a_b_c_d){//the next noise pixel is located at lower or upper pixel
            now_width += width-1;
            if(max_a_b == a){
                now_height -= 1;
            }else{
                now_height += 1;
            }
        }else{//the next noise pixel is located at right lower or upper right pixel
            now_width += width;
            if(max_c_d == c){
                now_height -= 1;
            }else{
                now_height += 1;
            }
        }
    }
}

File: weibo.cn/python/spliter/spliter.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
from __future__ import print_function
from __future__ import division

import os
import sys

import cv2
import numpy as np
import uuid

spliter_dir =os.path.dirname(os.path.abspath(__file__))
home_dir= os.path.dirname(spliter_dir)
sys.path.append(home_dir)
sys.path.append(spliter_dir)
from captcha_utils import CaptchaUtils
from common.common import IMAGE_HEIGHT, IMAGE_WIDTH

__all__ = ["Spliter"]


class Spliter:

    def __init__(self, save_dir):
        self.save_dir = save_dir
        if not os.path.isdir(save_dir):
            os.makedirs(save_dir)

    def split_letters(self, path):
        letters = [0]*4

        bare_name, ext = os.path.splitext(path)
        if ext == '.gif':
            from PIL import Image
            with Image.open(path) as im:
                path = bare_name+'.png'
                im.save(path)

        image = cv2.imread(path, cv2.IMREAD_COLOR)

        image = self.clear_noise(image)
        # cv2.imshow("clear", image)
        # cv2.waitKey(0)

        splits = [0]*8
        CaptchaUtils.vertical_project(image, splits)
        if splits[7] < image.shape[1]:  # assume the final split is in the range of image.
            for i in range(0, 8, 2):
                letters[i//2] = image[0:image.shape[0], splits[i]:splits[i+1], ].copy()

        return letters

    def split_and_save(self, filename):
        letters = self.split_letters(filename)
        for every_letter in letters:
            every_letter = self.format_splited_image(every_letter)
            self.__save_image(every_letter)

    def clear_noise(self, image):
        image = cv2.flip(image, -1,)
        clear_horizontal_noise_line(image)
        image = cv2.flip(image, -1, )
        clear_horizontal_noise_line(image)
        clear_color(image)

        image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
        image = cv2.threshold(image, 150, 255, cv2.THRESH_BINARY)[1]
        CaptchaUtils.clear_peper_noise(image, 2)

        return image

    def format_splited_image(self, splited_image):
        im_width = splited_image.shape[1]
        im_height = splited_image.shape[0]
        if im_width > IMAGE_WIDTH:
            return
        if im_height <= 0 or im_width <= 0:
            return

        out_width = IMAGE_WIDTH
        out_height = IMAGE_HEIGHT

        offset_x = int(abs(out_width - splited_image.shape[1]) / 2)
        offset_y = int(abs(out_height - splited_image.shape[0]) / 2)

        out = np.ones((32, 32)) * 255
        out[offset_y: offset_y+im_height, offset_x: offset_x+im_width] = splited_image

        return out

    def __save_image(self, formatted_image):
        if formatted_image is None:
            return
        fname = str(uuid.uuid4())
        path = os.path.join(self.save_dir, fname)+'.png'
        cv2.imwrite(path, formatted_image)


def is_black(i, j, image):
    if j < 0 or j > (image.shape[0] - 1):
        return False
    b = image[j, i][0]
    g = image[j, i][1]
    r = image[j, i][2]
    average = (int(r) + int(g) + int(b))/3
    if r < 244 and abs(average-b) < 4 and abs(average-g) < 4 and abs(average-r) < 4:
        return True
    return False


def clear_color(image):
    for i in range(image.shape[1]):
        for j in range(image.shape[0]):
            if is_black(i, j, image):
                image[j][i][0] = 20
                image[j][i][1] = 20
                image[j][i][2] = 20


def get_horizontal_noise_line_width(image, now_height, now_width):
    if now_height < 0 or now_height > (image.shape[0] - 1):
        return False
    end_width = now_width
    while end_width < image.shape[1] \
            and image[now_height][end_width][0] < 12 \
            and image[now_height][end_width][1] < 12 \
            and image[now_height][end_width][2] < 12:

        # print(image[now_height][end_width][0],
        #       image[now_height][end_width][1],
        #       image[now_height][end_width][2])

        end_width += 1

    return end_width - now_width


def clear_horizontal_noise_line(image):
    first_height = 0
    has_find = False
    for i in range(image.shape[0]):
        if image[i][0][0] < 12 and image[i][0][1] < 12 and image[i][0][2] < 12 \
                and get_horizontal_noise_line_width(image, i, 0) >= 2:
            first_height = i
            has_find = True
    # print("first:", first_height)

    if not has_find:
        return
    now_width = 0
    now_height = first_height
    while now_width < image.shape[1]:
        width = get_horizontal_noise_line_width(image, now_height, now_width)
        # print(now_width, now_height, "width", width)

        # clear the horizontal noise line
        for i in range(now_width, now_width+width-1):
            top_num = 0
            bottom_num = 0
            # the upper pixel
            if is_black(i, now_height-1, image):
                top_num += 1
            # the upper left pixel
            if is_black(i-1, now_height-1, image):
                top_num += 1
            # the upper right pixel
            if is_black(i+1, now_height-1, image):
                top_num += 1
            # the lower pixel
            if is_black(i, now_height+1, image):
                bottom_num += 1
            # the left lower pixel
            if is_black(i-1, now_height+1, image):
                bottom_num += 1
            # the right lower pixel
            if is_black(i+1, now_height+1, image):
                bottom_num += 1

            if now_height != 0 and now_height != image.shape[0]:
                if top_num > 0 and bottom_num > 0:
                    continue

            image[now_height][i][0] = 255
            image[now_height][i][1] = 255
            image[now_height][i][2] = 255

        # find the next noise pixel
        a = get_horizontal_noise_line_width(image, now_height - 1, now_width + width - 1)
        b = get_horizontal_noise_line_width(image, now_height + 1, now_width + width - 1)
        c = get_horizontal_noise_line_width(image, now_height - 1, now_width + width)
        d = get_horizontal_noise_line_width(image, now_height + 1, now_width + width)
        if now_height == 0:
            a = 0
            c = 0

        if now_height == (image.shape[0] - 1):
            b = 0
            d = 0

        max_a_b = max(a, b)
        max_c_d = max(c, d)
        max_a_b_c_d = max(max_a_b, max_c_d)
        if max_a_b_c_d < 2:
            break
        if max_a_b == max_a_b_c_d:
            now_width += width-1
            if max_a_b == a:
                now_height -= 1
            else:
                now_height += 1

        else:
            now_width += width
            if max_c_d == c:
                now_height -= 1
            else:
                now_height += 1

